libsmtutil/CHCSmtLib2Interface.cpp

```diff
@@ -223,3 +226,282 @@ std::string CHCSmtLib2Interface::createHeaderAndDeclarations() {
+					return std::make_shared<BitVectorSort>(std::stoul(asAtom(args[2])));
+			}
+			smtAssert(false, "Unknown sort encountered");
+		}
+
+		smtutil::Expression parseQuantifier(
+			std::string const& quantifierName,
+			std::vector<SMTLib2Expression> const& varList,
+			SMTLib2Expression const& coreExpression
+		)
+		{
+			std::vector<std::pair<std::string, SortPointer>> boundVariables;
+			for (auto const& sortedVar: varList)
+			{
+				smtAssert(!isAtom(sortedVar));
+				auto varSortPair = asSubExpressions(sortedVar);
+				smtAssert(varSortPair.size() == 2);
+				boundVariables.emplace_back(asAtom(varSortPair[0]), toSort(varSortPair[1]));
+			}
+			for (auto const& [var, sort] : boundVariables)
+			{
+				smtAssert(m_knownVariables.find(var) == m_knownVariables.end()); // TODO: deal with shadowing?
+				m_knownVariables.emplace(var, sort);
+			}
+			auto core = toSMTUtilExpression(coreExpression);
+			for (auto const& [var, sort] : boundVariables)
+			{
+				smtAssert(m_knownVariables.find(var) != m_knownVariables.end());
+				m_knownVariables.erase(var);
+			}
+			return Expression(quantifierName, {core}, SortProvider::boolSort); // TODO: what about the bound variables?
+
+		}
+
+		smtutil::Expression toSMTUtilExpression(SMTLib2Expression const& _expr)
+		{
+			return std::visit(
+				GenericVisitor{
+					[&](std::string const& _atom)
+					{
+						if (_atom == "true" || _atom == "false")
+							return smtutil::Expression(_atom == "true");
+						else if (isNumber(_atom))
+							return smtutil::Expression(_atom, {}, SortProvider::sintSort);
+						else if (auto it = m_knownVariables.find(_atom); it != m_knownVariables.end())
+							return smtutil::Expression(_atom, {}, it->second);
+						else // assume this is a predicate with sort bool; TODO: Context should be aware of predicates!
+							return smtutil::Expression(_atom, {}, SortProvider::boolSort);
+					},
+					[&](std::vector<SMTLib2Expression> const& _subExpr)
+					{
+						SortPointer sort;
+						std::vector<smtutil::Expression> arguments;
+						if (isAtom(_subExpr.front()))
+						{
+							std::string const& op = asAtom(_subExpr.front());
+							if (op == "!")
+							{
+								// named term, we ignore the name
+								smtAssert(_subExpr.size() > 2);
+								return toSMTUtilExpression(_subExpr[1]);
+							}
+							if (op == "exists" || op == "forall")
+							{
+								smtAssert(_subExpr.size() == 3);
+								smtAssert(!isAtom(_subExpr[1]));
+								return parseQuantifier(op, asSubExpressions(_subExpr[1]), _subExpr[2]);
+							}
+							for (size_t i = 1; i < _subExpr.size(); i++)
+								arguments.emplace_back(toSMTUtilExpression(_subExpr[i]));
+							if (auto tupleSort = lookupKnownTupleSort(op); tupleSort)
+							{
+								auto sortSort = std::make_shared<SortSort>(tupleSort.value());
+								return Expression::tuple_constructor(Expression(sortSort), arguments);
+							}
+							if (auto it = m_knownVariables.find(op); it != m_knownVariables.end())
```

Comment:
As a side note for future reference (don't bother changing it now), we have `util::contains(m_knownVariables, var)` as an alternative - eventually, once we switch over to c++20, we'll get the `.contains()` member for all `std::` associative containers, so we there won't be need for any of these monstrosities :)

Reply:
It looks like `util::contains` does not work for `std::map`.
It uses `std::find`, but that is too general, it just iterates over the container and uses `==` to compare against the value. But that does not work for map, because dereferencing iterator gives you key-value pair, but I only want to find by key.

Reply:
Ah, you're right, there's no specialization for `map`.